rts/motoko-rts/src/bitmap.rs

```diff
@@ -0,0 +1,116 @@
+    // Also round allocation up to 8-bytes to make iteration efficient. We want to be able to read
+    // 64 bits in a single read and check as many bits as possible with a single `word != 0`.
+    let bitmap_bytes = Bytes(((bitmap_bytes + 7) / 8) * 8);
+    // Allocating an actual object here as otherwise dump_heap gets confused
+    let blob = alloc_blob(bitmap_bytes).unskew() as *mut Blob;
+    memzero(blob.payload_addr() as usize, bitmap_bytes.to_words());
+
+    BITMAP_PTR = blob.payload_addr()
+}
+
+pub unsafe fn free_bitmap() {
+    BITMAP_PTR = core::ptr::null_mut();
+}
+
+pub unsafe fn get_bit(idx: u32) -> bool {
+    let byte_idx = idx / 8;
+    let byte = *BITMAP_PTR.add(byte_idx as usize);
+    let bit_idx = idx % 8;
+    (byte >> bit_idx) & 0b1 == 0b1
+}
+
+pub unsafe fn set_bit(idx: u32) {
+    let byte_idx = idx / 8;
+    let byte = *BITMAP_PTR.add(byte_idx as usize);
+    let bit_idx = idx % 8;
+    let new_byte = byte | (0b1 << bit_idx);
+    *BITMAP_PTR.add(byte_idx as usize) = new_byte;
+}
+
+pub struct BitmapIter {
+    /// Size of the bitmap, in 64-bit words words. Does not change after initialization.
+    size: u32,
+    /// Current 64-bit word index
+    current_word_idx: u32,
+    /// Current 64-bit word in the bitmap that we're iterating. We read in 64-bit chunks to be able
+    /// to check as many bits as possible with a single `word != 0`.
+    current_word: u64,
+    /// Bits left in the current 64-bit word. Used to compute index of a bit in the bitmap. We
+    /// can't use a global index here as we don't know how much to bump it when `current_word` is
+    /// 0 and we move to the next 64-bit word.
+    bits_left: u32,
+}
+
+pub unsafe fn iter_bits() -> BitmapIter {
+    let blob_len_bytes = (BITMAP_PTR.sub(size_of::<Blob>().to_bytes().0 as usize) as *mut Obj)
+        .as_blob()
+        .len()
+        .0;
+
+    debug_assert_eq!(blob_len_bytes % 8, 0);
+
+    let blob_len_64bit_words = blob_len_bytes / 8;
+
+    let current_word = if blob_len_64bit_words == 0 {
+        0
+    } else {
+        *(BITMAP_PTR as *const u64)
+    };
+
+    BitmapIter {
+        size: blob_len_64bit_words,
+        current_word_idx: 0,
+        current_word,
+        bits_left: 64,
+    }
+}
+
+impl Iterator for BitmapIter {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<u32> {
+        debug_assert!(self.current_word_idx <= self.size);
+
+        // Outer loop iterates 64-bit words
+        loop {
+            if self.current_word == 0 && self.current_word_idx == self.size {
+                return None;
+            }
+
+            // Inner loop iterates bits in the current word
+            while self.current_word != 0 {
+                if self.current_word & 0b1 == 0b1 {
+                    let bit_idx = (self.current_word_idx * 64) + (64 - self.bits_left);
+                    self.current_word >>= 1;
+                    self.bits_left -= 1;
+                    return Some(bit_idx);
```

Comment:
Is this Some allocation cheap in Rust or would it be worth avoiding iterators altogether for this low-level code.

Reply:
`Option<u32>` values are represented as two `u32`s on Wasm, one as tag (indicating whether the value is `None` or `Some(...)`) and one for the value of `Some`. Here's an example function that returns a `Option<u32>`: https://godbolt.org/z/eYT6oMhhf it takes two arguments, one for the address of the resulting `Option<u32>`, one for the argument in the source code (the `u32`). It writes `1` to the first memory location (so tag = `Some`), and the argument to the second memory location.

In our case, heap is 4GiB and each word (32-bit) gets a bit, so we need `1024 * 1024 * 1024` bits in the worst case (actually less than that, as we can't allocate a bitmap if the heap is full). So we could use one larger than that to indicate the `None` case here.

I will implement this and update the benchmark results.

Reply:
> In our case, heap is 4GiB and each word (32-bit) gets a bit, so we need 1024 * 1024 * 1024 bits in the worst case (actually less than that, as we can't allocate a bitmap if the heap is full). So we could use one larger than that to indicate the None case here.

This made 2.7% reduction in instructions in some cases so I will commit my code that implements this now.

Reply:
Nice catch!

Reply:
Done.

Reply:
Cool!